lib/Model/ConvertFunctionType.cpp

```diff
@@ -0,0 +1,664 @@
+        Result[Argument.Index] = Distributed;
+        Counter = NextIndex;
+      } else {
+        const size_t Limit = CC::MaximumGPRsPerAggregateArgument;
+
+        auto [Distributed, NextIndex] = considerRegisters<CC>(*MaybeSize,
+                                                              Limit,
+                                                              Counter,
+                                                              Registers,
+                                                              SplitArgs);
+        Result[Argument.Index] = Distributed;
+        Counter = NextIndex;
+      }
+    }
+  }
+
+  return Result;
+}
+
+template<typename CallConv>
+static std::optional<DistributedArguments>
+distributeArguments(const SortedVector<model::Argument> &Arguments) {
+  if constexpr (CallConv::ArgumentsArePositionBased)
+    return distributeArgumentsBasedOnPositions<CallConv>(Arguments);
+  else
+    return distributeArgumentsFreely<CallConv>(Arguments);
+}
+
+template<typename CallConv>
+static std::optional<DistributedArgument>
+distributeReturnValue(const model::QualifiedType &ReturnValueType) {
+  auto MaybeSize = ReturnValueType.size();
+  if (MaybeSize == std::nullopt)
+    return std::nullopt;
+
+  using CC = CallConv;
+  if (ReturnValueType.isFloat()) {
+    const auto &Registers = CC::VectorReturnValueRegisters;
+    return considerRegisters<CC>(*MaybeSize, 1, 0, Registers, false).first;
+  } else {
+    const auto &Registers = CC::GeneralPurposeReturnValueRegisters;
+    if (ReturnValueType.isScalar()) {
+      const size_t L = CC::MaximumGPRsPerScalarReturnValue;
+      return considerRegisters<CC>(*MaybeSize, L, 0, Registers, false).first;
+    } else {
+      const size_t L = CC::MaximumGPRsPerAggregateReturnValue;
+      return considerRegisters<CC>(*MaybeSize, L, 0, Registers, false).first;
+    }
+  }
+}
+
+static model::QualifiedType
+chooseArgumentType(const model::QualifiedType &ArgumentType,
+                   model::Register::Values Register,
+                   const RegisterList &RegisterList,
+                   model::Binary &TheBinary) {
+  if (RegisterList.size() > 1) {
+    return buildGenericType(Register, TheBinary);
+  } else {
+    auto ResultType = ArgumentType;
+    auto MaybeSize = ArgumentType.size();
+    auto TargetSize = model::Register::getSize(Register);
+
+    if (!MaybeSize) {
+      return buildType(Register, TheBinary);
+    } else if (*MaybeSize > TargetSize) {
+      auto Qualifier = model::Qualifier::createPointer(TargetSize);
+      ResultType.Qualifiers.emplace_back(Qualifier);
+    } else if (!ResultType.isScalar()) {
+      return buildGenericType(Register, TheBinary);
+    }
+
+    return ResultType;
+  }
+}
+
+template<model::abi::Values A>
+std::optional<model::RawFunctionType>
+convertToRawFunctionTypeImpl(const model::CABIFunctionType &Function,
+                             model::Binary &TheBinary) {
+  using CallingConvention = ::abi::CallingConventionTrait<A>;
```

Comment:
You seem to be using functions templated with similar arguments a lot. Consider grouping them in a template `struct` with static functions. It will make a lot of `<CallingConvention>` disappear.

Reply:
I optimized the parameters a bit (see https://github.com/revng/revng/pull/98/commits/d4c791dd5b4968092a98057d010522474f7c36c0 and https://github.com/revng/revng/pull/98/commits/c94727e1f3b8bae8c0d86a66b44e136d8c445700). Do you still think I should go for a struct?

Reply:
Uhm, yes. But I'll leave that to you.

Reply:
Done.